File: frontend/src/api/knowledge.ts
import { request, sendFile } from "./http";

export type Material = {
  id: string;
  filename: string;
  source: string;
  mime: string;
  status: string;
  error?: string | null;
  size_bytes: number;
  chunk_count: number;
  created_at: string;
};

export type RecallHit = {
  chunk_id: string;
  material_id: string;
  filename: string;
  ordinal: number;
  score: number;
  text: string;
};

export const knowledgeApi = {
  materials: () => request<Material[]>("/api/admin/materials"),
  uploadMaterial: (file: File) => sendFile<Material>("/api/admin/materials", file),
  deleteMaterial: (id: string) => request<{ ok: string }>(`/api/admin/materials/${id}`, { method: "DELETE" }),
  recall: (query: string) =>
    request<{ query: string; hits: RecallHit[]; index_status: string }>("/api/admin/recall", {
      method: "POST",
      body: JSON.stringify({ query }),
    }),
};

File: backend/app/schemas/knowledge.py
"""Request and response shapes for materials and recall."""

from datetime import datetime

from pydantic import BaseModel, Field


class MaterialOut(BaseModel):
    id: str
    filename: str
    source: str
    mime: str
    status: str
    error: str | None = None
    size_bytes: int
    chunk_count: int
    created_at: datetime


class ChunkOut(BaseModel):
    id: str
    ordinal: int
    text: str
    token_estimate: int


class MaterialDetail(MaterialOut):
    chunks: list[ChunkOut] = Field(default_factory=list)


class RecallHit(BaseModel):
    chunk_id: str
    material_id: str
    filename: str
    ordinal: int
    score: float
    text: str


class RecallOut(BaseModel):
    query: str
    hits: list[RecallHit]
    index_status: str
